File: nodejs/patterns/src/creational/lazy_initialization.js
class LazyDataset {
  constructor() {
    this.records = null;
    this.loadCount = 0;
  }

  getRecords() {
    if (!this.records) {
      this.loadCount += 1;
      this.records = ["row-1", "row-2", "row-3"];
    }
    return this.records;
  }
}

function demo() {
  const dataset = new LazyDataset();
  dataset.getRecords();
  dataset.getRecords();
  return dataset.loadCount === 1;
}

module.exports = { LazyDataset, demo };

File: python/patterns/src/prototype_patterns/creational/lazy_initialization.py
"""Lazy Initialization pattern example."""

from __future__ import annotations


class LazyDataset:
    def __init__(self) -> None:
        self._records: list[str] | None = None
        self.load_count = 0

    @property
    def records(self) -> list[str]:
        if self._records is None:
            self.load_count += 1
            self._records = ["row-1", "row-2", "row-3"]
        return self._records


def demo() -> bool:
    dataset = LazyDataset()
    first = dataset.records
    second = dataset.records
    return first == second and dataset.load_count == 1
